Fix transition jar file references issue

derheld42 opened the pull request:
Handles the case where:
Java Project A outputs a.jar containing package A
OSGI Project B has a version=file reference to a.jar and includes package A in export-packages.
OSGI Project C has buildpath reference to B and uses something in package A.

Without this change, Project C couldn't build due to
"not accessible due to restriction on required project" errors

Signed-off-by: Carter Smithhart <carter.smithhart@gmail.com>

derheld42 commented:
This allows me to close out #420

bjhargrave commented:
How does this match the offline build behavior? Does this improve the fidelity?

derheld42 commented:
I just tweaked the implementation. I forget to verify that non-exported packages shouldn't be accessible. I just tested that with the tweak.

I believe this brings bndtools closer to matching bnd. It definitely addresses (part of?) a problem with an eclipse project wrapping a .jar and providing portions of that jar as exported packages.

bjhargrave commented on bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java:
```diff
@@ -316,6 +324,27 @@ static boolean isProjectContainer(Container container) {
+        } catch (IOException e) {}
+        if (tmp != null && tmp.size() != 0) {
+            tmp.add(JavaCore.newAccessRule(new Path("**"), IAccessRule.K_NON_ACCESSIBLE));
+            return tmp.toArray(new IAccessRule[tmp.size()]);
```
There is duplicate code here (add ** rule, convert to array). It would seem that it is better move that common code to the end of the method so all exit paths end here, no? That is, this method needs more refactoring for your change.

derheld42 commented:
I tested another case:
Project A uses jar file B with packace C
Project A also overrides a class in package C
Project A exports package C
Project D has a buildpath reference to A.
Project D does sees the updated class from Project A.

The only thing that didn't work as expected was that I had to do a "rebuild" on project A to get the classpath to get recalculated so that Project D saw the right version of package C. This happens if package C was just created in Project A.

I think eclipse is providing the jar B package C instead of Project A package C. A rebuild cleared that up.

I'm sure there's a fix for to make it work right -- although I don't know if this is a likely case people will run into.

The basic problem is fixed with this pull request. There may be strange cases like this test right here that don't work "perfectly"

derheld42 commented:
Refactored and rebased.

I ran into nastiness that slowed me down related to .gitattributes *.css and *.properties. I had to comment those 2 lines out. Otherwise everytime I did a git clean or git checkout, it would think I changed these two:
	modified:   bndtools.core/resources/unprocessed/intro/css/whatsnew.css
	modified:   bndtools.core/resources/unprocessed/intro/css/whatsnew.properties

Although now that I set core.autocrlf=false, it seems to have gone away. ugg.
I'm using git version 1.9.1 on ubuntu.

m4rr5 commented:
Did you by any chance also validate if this PR fixes #937 as it sounds related.

bjhargrave commented on bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java:
```diff
@@ -303,16 +311,50 @@ static boolean isProjectContainer(Container container) {
+            Packages exportPkgs = model.getExports();
+            ZipFile zip = null;
+            try {
+                zip = new ZipFile(c.getFile());
```
Instead of using ZipFile, I think it would be better to use aQute.bnd.osgi.Jar. You can then use the Jar object to interrogate the directories or package (which ever is the correct choice here).

bjhargrave commented:
> I ran into nastiness that slowed me down related to .gitattributes *.css and *.properties. I had to comment those 2 lines out. Otherwise everytime I did a git clean or git checkout, it would think I changed these two:
> modified: bndtools.core/resources/unprocessed/intro/css/whatsnew.css
> modified: bndtools.core/resources/unprocessed/intro/css/whatsnew.properties
> 
> Although now that I set core.autocrlf=false, it seems to have gone away. ugg.
> I'm using git version 1.9.1 on ubuntu.

I'll release a PR to fix this.

bjhargrave commented:
> I'll release a PR to fix this.

That issue is fixed by https://github.com/bndtools/bndtools/pull/1001.

derheld42 commented:
I wonder if there's any good way to set up eclipse test projects for exercising weird stuff like this. hmm.

derheld42 commented:
This last commit has a bit longer explanation for what this commit is now all about.

I now actually believe this will increase the fidelity between bndtools and bnd in terms of no-transitive dependencies. I think there might be a fringe case or 2 left, but this should resolve some of the other outstanding issues related to access issues and version=file references.

Outside of this issue, the NewBuilder doesn't seem to kick off a build when the classpath changes sometimes... and I don't know why. I traced through the NewBuilder code and it correctly updates the classpath, then returns a dependent project like it does elsewhere in the class. But a project rebuild is then required to get a project to build without errors. I have an example test case above in this issue if someone is interested.

bjhargrave commented on bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java:
```diff
@@ -255,26 +273,59 @@ public static void setClasspathEntries(IJavaProject javaProject, IClasspathEntry
+                System.out.println("--- " + f);
+            }
+        }
+
```
I assume the above code should not be part of the final PR?

derheld42 replied:
I thought it was useful for debugging purposes. I keep adding code like that to aid debugging. Might as well add it to the baseline.

bjhargrave replied:
My opinion would be to leave it out.

derheld42 replied:
I switched it to a logger in the updated commit.

bjhargrave commented on bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java:
```diff
@@ -210,17 +214,31 @@ public static void setClasspathEntries(IJavaProject javaProject, IClasspathEntry
+                calculateWorkspaceBundleAccessRules(projectAccessRulesExports, projectAccessRulesPrivates, c, model);
+            }
+        }
+        //[cs] Add ** NO access to end of rule sets that have rules.
```
Why is there "[cs]" in front of all the comments? git blame can tell us you write these comments.

derheld42 replied:
Habit of mine. It's a quick way of knowing who added the comment. I find it useful. If the bndtools coding standard says don't do that, I can remove it.

bjhargrave commented on bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java:
```diff
@@ -342,12 +416,26 @@ static void calculateWorkspaceBundleAccessRules(Map<Project,List<IAccessRule>> p
+                        exports.add(JavaCore.newAccessRule(new Path(pathStr), IAccessRule.K_ACCESSIBLE));
+                    }
+                }
+                //TODO -- fix bnd bug.
```
What is the bug here?

derheld42 replied:
I think so. Looks like copy and paste bug. The containsbinaryname method seems wrong. I didn't write anything up yet.

derheld42 replied:
@pkriens is there a copy 'n paste bug in aQute.bnd.osgi.Packages.containsBinaryName ? It has the same implementation as containsFQN()

bjhargrave replied:
Why don't you open a bnd issue for this? It can be tracked separately.

derheld42 replied:
Done. bndtools/bnd#782 written.

derheld42 commented:
Any concerns about merging this in? It should see some bake time in the baseline before another release.

I know I have the power to merge it in myself, but I'm not quite sure when a review must/should happen... What's our protocol?

I suppose a "bad" change could always get reverted... but is that how we're working?

pkriens commented:
It is rather simple :-) 

If you make a simple fix, merge it yourself. If it is in any way imaginable contentious (workspace layout for example :-), you feel unsure, or it is large then ping the group and ask for review. As you said, it is easy to undo.

In this case, go for it. Looks like you did some serious work in this black art part of bndtools ...

derheld42 commented:
doh... I missed something. I'm either going to fix it or revert it.

pkriens commented:
:-)
